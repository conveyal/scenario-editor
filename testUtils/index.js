export function mockComponents (components) {
  if (Array.isArray(components)) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
