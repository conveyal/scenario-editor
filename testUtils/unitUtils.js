/* globals describe, expect, it */

import React from 'react'
import renderer from 'react-test-renderer'

/**
 * Performs a basic render test of a Component
 *
 * @param  {Mixed} components Can be either an array of objects or a single object
 *   Each object can have the following properties:
 *   - children: Any children to render within the component
 *   - component (required): The component to render
 *   - name: How to classify the test in the `describe section`
 *   - notToBeCalledFns: An array of jest.fn()'s that will be asserted as not called during basic render test
 *   - props (required): The props of the component to render
 *   - toBeCalledFns: An array of jest.fn()'s that will be asserted as called during basic render test
 */
export function basicRenderTest (components) {
  // prepare components to be tested
  if (!Array.isArray(components)) {
    components = [components]
  }

  // run tests on each component
  components.forEach((component) => {
    const Component = component.component

    describe(`Component > ${component.name}`, () => {
      it('renders correctly', () => {
        const tree = renderer.create(
          <Component
            {...component.props}
            >
            {component.children}
          </Component>
        ).toJSON()
        expect(tree).toMatchSnapshot()

        if (Array.isArray(component.toBeCalledFns)) {
          component.toBeCalledFns.forEach((fn) => {
            expect(fn).toBeCalled()
          })
        }

        if (Array.isArray(component.notToBeCalledFns)) {
          component.notToBeCalledFns.forEach((fn) => {
            expect(fn).not.toBeCalled()
          })
        }
      })
    })
  })
}
