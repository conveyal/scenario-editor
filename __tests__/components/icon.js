import { basicRenderTest } from '../../testUtils/unitUtils'

import Icon from '../../lib/components/icon'

basicRenderTest({
  component: Icon,
  name: 'Icon',
  props: {
    type: 'pencil',
    className: 'test'
  }
})
