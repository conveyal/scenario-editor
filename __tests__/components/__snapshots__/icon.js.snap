exports[`Component > Icon renders correctly 1`] = `
<i
  className="fa fa-pencil fa-fw test" />
`;
